rngd.c

```diff
@@ -752,46 +752,80 @@ static error_t parse_opt (int key, char *arg, struct argp_state *state)
 static struct argp argp = { options, parse_opt, NULL, doc };
 
 
-static int update_kernel_random(struct rng *rng, int random_step,
-	unsigned char *buf, fips_ctx_t *fipsctx_in)
+static int update_kernel_random(int random_step)
 {
-	unsigned char *p;
-	int fips = 0;
+	unsigned char buf[FIPS_RNG_BUFFER_SIZE]; /* random_step was checked to be <= FIPS_RNG_BUFFER_SIZE */
 	int rc;
+	struct rng *iter;
 
-	if (!arguments->ignorefail)
-		fips = fips_run_rng_test(fipsctx_in, buf);
-	if (fips && !arguments->ignorefail)
-		return 1;
+	message(LOG_DAEMON|LOG_DEBUG, "entropy successfully gathered, preparing it for the kernel\n");
 
-	for (p = buf; p + random_step <= &buf[FIPS_RNG_BUFFER_SIZE];
-		 p += random_step) {
+	while(true)
+	{
 		if (!server_running)
 			return 0;
 		if (do_reseed) {
 			do_reseed = false;
 			alarm(arguments->force_reseed);
 		}
-		rc = random_add_entropy(p, random_step);
+
+		/* mix the sources on byte-level: ensure we always feed data from all available sources to the kernel
+		 * helps to mitigate problems should a source not be as random as expected */
+		int p = 0;
+		while(p < random_step) {
+			int progress = p;
+			for (int i = 0; i < ENT_MAX; ++i) {
+				iter = &entropy_sources[i];
+				if (!iter->entropy_buf.valid)
+					continue;
+
+				buf[p++] = iter->entropy_buf.entropy[iter->entropy_buf.used_pos++];
+
+				if (iter->entropy_buf.used_pos == FIPS_RNG_BUFFER_SIZE)
+					iter->entropy_buf.valid = false;
+
+				if(p >= random_step)
+					break;
+			}
+
+			/* abort data preparation when no data was added to the buffer in one loop = no valid sources left
+			 * this wastes a few bytes when FIPS_RNG_BUFFER_SIZE is not a multiple of random_step
+			 * but it makes the logic easier to implement and read */
+			if (p == progress)
+				return 0;
+		}
+
+		rc = random_add_entropy(buf, random_step);
 		if (rc == -1)
-			return 1;
+		{
```

Comment:
style nit: brace should follow the conditional, rather than be on a line by itself please

Reply:
ah, the braces on the next line is the style i normally use. i tried to adapt to the rng-tools style, but it seems like i overlooked some cases.

I will update the pull request.

Reply:
fixed in new revision